### ISSUE TITLE
Issues found with syntax hightlighting

### ISSUE THREAD

**Issue by stsewd** [nvim-treesitter/nvim-treesitter#78]
I have been using the highlight feature for a couple of weeks now (python mainly).

There are three issues I have found so far:

- ~While typing some syntax may be invalid, and everything goes red~ (fixed with `highlight link TSError Normal`)

![red3](https://user-images.githubusercontent.com/4975310/84316280-7c468e00-ab30-11ea-913d-e8e73d76f827.png)
![red2](https://user-images.githubusercontent.com/4975310/84316277-7badf780-ab30-11ea-826a-4bcc05c0e2bc.png)

- ~Sometimes there is an error (not sure how to replicate), and the text gets replaced with an error from lua. This is fixed by reloading the text with `:e`.~ Fixed, haven't experienced this anymore.

![luaerrro](https://user-images.githubusercontent.com/4975310/84420183-4b716200-abdf-11ea-9d79-94e5ff99ff2a.png)

- When the using folds, some parts of the code doesn't get highlighted. This happens when I open a file where the text is folded (syntax is highlighted as normal), then go to the last line with `G` (syntax isn't highlighted, the whole text is white). This is fixed by forcing a redraw with `<C-L>`.

**Comment by theHamsta** [comment]
Hi, the ERROR highlight is easily fixable:

```lua
    require "nvim-treesitter".setup()
    require "nvim-treesitter.highlight"
    local hlmap = vim.treesitter.TSHighlighter.hl_map

    --Misc
    hlmap.error = nil
```

We probably should default to that.

The issues that you're observing are problems introduced by the last runtime update of Neovim built-in treesitter. Apparently, the code that's performing the regex queries still has some bugs.

**Comment by steelsojka** [comment]
Would this be causing this issue I'm seeing?

![Screen Shot 2020-07-09 at 7 23 29 AM](https://user-images.githubusercontent.com/2062154/87041231-8f905c00-c1b7-11ea-91a4-1d655ed626f7.png)

**Comment by vigoux** [comment]
Yep, both of them are related, the way you can help is tell us the edit you did to cause that. Because that's actually the tree getting out of sync.

**Comment by steelsojka** [comment]
@vigoux Sounds good. I'm getting it a lot on typescript code. Seems to be on just inserting text. Is this something we can fix or does it need to be fixed upstream in nvim?

**Comment by vigoux** [comment]
It needs to be fixed upstream. And it is related to especially gnarly parts of neovim, thus it requires quite q bit of effort (at least for me...) to fix.

**Comment by glepnir** [comment]
Same issue here.  I found it has some problems with syntax.  I defined some syntax group in my colorscheme. when i install treesitter. It didn't follow syntax group,It even changed the color scheme.

before:
![3](https://user-images.githubusercontent.com/41671631/87790570-54ef8900-c873-11ea-8bab-43f08e1b3788.png)


after:
![1](https://user-images.githubusercontent.com/41671631/87790562-502ad500-c873-11ea-8a20-8f21ae16b1f3.png)

With typescript. It works better than default `regexengine2` ,But it looks like I didn't get more color schemes.

before:
![4](https://user-images.githubusercontent.com/41671631/87791037-2aea9680-c874-11ea-8402-3f9c0073ff21.png)

after:
![2](https://user-images.githubusercontent.com/41671631/87791042-2de58700-c874-11ea-8a21-da3c6378712d.png)

**Comment by vigoux** [comment]
@glepnir 

Treesitter highlighting is a completely different mechanism than standard highlighting. tbh they don't even share any single line of code (on the implementation side). Thus if you define syntax groups, treesitter will just not know that, you have to use queries to do so.

The colors changed and that's normal, treesitter highlighting tries to bring meaning together with colors, here for example, in red you find keywords, green are function calls and so on. So the colorscheme changed yes, but if you compare you go code and you typescript code, they _look_ uniform and consistent on the highlighting side. To customize the behaviour though, you can customize the highlight groups `nvim-treesitter` uses using `:highlight link {TSHlGroup} {TheGroupYouWant}`.

**Comment by glepnir** [comment]
@vigoux  Aha thanks. About this `highlight link {TSHlGroup} {TheGroupYouWant}`. I think  i know the `{TheGroupYouWant}` that the syntax group i defined in my colorscheme of ts or other language syntax group.  This ` {TSHlGroup}` mean these?

```lua
  require "nvim-treesitter.highlight"
    local hlmap = vim.treesitter.TSHighlighter.hl_map

    --Misc
    hlmap.error = nil
    hlmap["punctuation.delimiter"] = "Delimiter"
    hlmap["punctuation.bracket"] = "Delimiter"

    -- Constants
    hlmap["constant"] = "Constant"
    hlmap["constant.builtin"] = "Type"
    hlmap["constant.macro"] = "Define"
    hlmap["string"] = "String"
    hlmap["string.regex"] = "String"
    hlmap["string.escape"] = "SpecialChar"
    hlmap["character"] = "Character"
    hlmap["number"] = "Number"
    hlmap["boolean"] = "Boolean"
    hlmap["float"] = "Float"

    -- Functions
    hlmap["function"] = "Function"
    hlmap["function.builtin"] = "Special"
    hlmap["function.macro"] = "Macro"
    hlmap["parameter"] = "Identifier"
    hlmap["method"] = "Function"
    hlmap["field"] = "Identifier"
    hlmap["property"] = "Identifier"
    hlmap["constructor"] = "Type"

    -- Keywords
    hlmap["conditional"] = "Conditional"
    hlmap["repeat"] = "Repeat"
    hlmap["label"] = "Label"
    hlmap["operator"] = "Operator"
    hlmap["keyword"] = "Repeat"
    hlmap["exception"] = "Exception"
    hlmap["include"] = "Include"
    hlmap["type"] = "Type"
    hlmap["type.builtin"] = "Type"
    hlmap["structure"] = "Structure"
```

**Comment by vigoux** [comment]
Hmm you have a pretty old version of the plugin, I mean [these](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/plugin/nvim-treesitter.vim#L22). Update to latest master (on both neovim and the plugin) so that you can configure this how you like it.

**Comment by glepnir** [comment]
Ok. I got it. I will try to define them. Thanks .

**Comment by PitcherTear22** [comment]
@glepnir which statusline is that?

**Comment by glepnir** [comment]
@PitcherTear22   It's my plugin [spaceline](https://github.com/hardcoreplayers/spaceline.vim)

**Comment by stsewd** [comment]
Just an update here, the only problem that isn't fixed is

> When the using folds, some parts of the code doesn't get highlighted. This happens when I open a file where the text is folded (syntax is highlighted as normal), then go to the last line with G (syntax isn't highlighted, the whole text is white). This is fixed by forcing a redraw with <C-L>.

**Comment by stsewd** [comment]
I can't replicate this anymore with folding, but it still happens with other cases like https://github.com/nvim-treesitter/nvim-treesitter/issues/235. I'm closing this in favor of the other issue.

**Comment by jacobduncan00** [comment]
> Same issue here. I found it has some problems with syntax. I defined some syntax group in my colorscheme. when i install treesitter. It didn't follow syntax group,It even changed the color scheme.
> 
> before: ![3](https://user-images.githubusercontent.com/41671631/87790570-54ef8900-c873-11ea-8bab-43f08e1b3788.png)
> 
> after: ![1](https://user-images.githubusercontent.com/41671631/87790562-502ad500-c873-11ea-8a20-8f21ae16b1f3.png)
> 
> With typescript. It works better than default `regexengine2` ,But it looks like I didn't get more color schemes.
> 
> before: ![4](https://user-images.githubusercontent.com/41671631/87791037-2aea9680-c874-11ea-8402-3f9c0073ff21.png)
> 
> after: ![2](https://user-images.githubusercontent.com/41671631/87791042-2de58700-c874-11ea-8a21-da3c6378712d.png)

What colorscheme and font is this?
